qcschema/dev/basis.py

```diff
@@ -0,0 +1,46 @@
+"""
+The json-schema for the Basis Set definition
+"""
+
+basis = {
+    "$schema": "http://json-schema.org/draft-04/schema#",
+    "name": "qcschema_basis",
+    "version": "1.dev",
+    "description": "The MolSSI Quantum Chemistry Basis Set Schema",
+    "type": "object",
+    "additionalProperties": False,
+    "properties": {
+        "schema_name": {
+            "type": "string",
+            "pattern": "^(qcschema_basis)$"
+        },
+        "schema_version": {
+            "type": "integer"
+        },
+        "description": {
+            "description": "Brief description of the basis set",
+            "type": "string"
+        },
+        "element_basis": {
+            "description": "Per-element basis data",
+            "type": "object",
+            "additionalProperties": False,
+            "patternProperties": {
+                "^\\d+$": {
+                    "$ref": "#/definitions/center_basis"
+                }
+            }
+        },
+        "atom_basis": {
+            "description":
+            "Basis set overrides for particular atoms or centers",
+            "type": "object",
+            "additionalProperties": False,
+            "patternProperties": {
+                "^\\d+$": {
+                    "$ref": "#/definitions/center_basis"
```

Comment:
I'm concerned about `atom_basis` and `element_basis` in that this (default by element and override by atom) is a very "input file" of specifying a basis attached to a molecule. In particular (not for the H2O-Zr), the order applied matters. And the user could be overriding in atom_basis by atom label, not center number. Would it make sense to consolidate element/atom_basis and have a list of center numbers as keys to the center_basis values. or dummy hash strings for the values and a separate dictionary with each atom `range(nat)` as keys and hashes as values?

Reply:
It depends on if we expect users to create theses schema by hand (which I would assume should never happen). If so, then just removing `element_basis` might be sufficient, and then converting `atom_basis` to a list. That gets unwieldy, though.

I kind of like the consolidated representation that you mention. I think I would prefer a dictionary with dummy keys, and then an array of len(nat) that points into that dict (this would eliminate 0 vs. 1-based indexing).

Anyone have other thoughts to this approach?

Reply:
The `Molecule` schema could be created by hand but in practice is nearly always fn-generated. I don't think practicality of user-generated `BasisSet` is something to optimize for, as one really wants a validation step in there.

An `atom_basis` list would be my first choice, but I known ppl are going to be protesting with good cause the wasted memory in a buckyball spec.

If you like it, too, then I'm all for the consolidated repr. And the list of basis hash keys melds nicely with all the arrays in `Molecule`. Hashes are also, incidentally, how I evaluate the effect of basis set on molecular symmetry in psi, and it works well (https://github.com/psi4/psi4/blob/bbc46d48a3565bb553f06d2365d21b290972098a/psi4/driver/qcdb/libmintsbasisset.py#L915).

Reply:
Updated the PR with this recommendation (and others). See if that is what you were envisioning